src/python/pants/jvm/compile.py

```diff
@@ -89,10 +124,19 @@ def for_targets(
         request types which are marked `root_only`.
         """
 
+        impls = jvm_request_types.classpath_entry_requests
+        usable_generators = jvm_request_types.code_generator_requests
+
+        # TODO: filter usable generators by acceptable languages
```

Comment:
By moving into `def classify_impl`, presumably?

Reply:
At this point, all of the generators in `usable_generators` will produce code that can be used in a JVM language. This is where we need a tie-breaking field on the sources to specify which language to use. I might add a `jvm_language` field, since it's really only acceptable to generate one language?

Reply:
E.g., `src/protobuf/some/proto@generator=java`

Reply:
@tdyas I suspect we want to restrict to one language per compiled artifact, until we have a good understanding of how code generated from the same sources to different languages can live together in a package artifact; if you've got a better idea of design here, I'm happy to talk about it in slack

Reply:
Yes, that was in the parametrization design as "dynamic parametrization".

Reply:
Punting, pending design work.